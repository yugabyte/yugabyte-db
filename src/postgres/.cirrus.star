"""Additional CI configuration, using the starlark language. See
https://cirrus-ci.org/guide/programming-tasks/#introduction-into-starlark

See also the starlark specification at
https://github.com/bazelbuild/starlark/blob/master/spec.md

See also .cirrus.yml and src/tools/ci/README
"""

load("cirrus", "env", "fs")


def main():
    """The main function is executed by cirrus-ci after loading .cirrus.yml and can
    extend the CI definition further.

    As documented in .cirrus.yml, the final CI configuration is composed of

    1) the contents of .cirrus.yml

    2) if defined, the contents of the file referenced by the, repository
       level, REPO_CI_CONFIG_GIT_URL variable (see
       https://cirrus-ci.org/guide/programming-tasks/#fs for the accepted
       format)

    3) .cirrus.tasks.yml
    """

    output = ""

    # 1) is evaluated implicitly

    # Add 2)
    repo_config_url = env.get("REPO_CI_CONFIG_GIT_URL")
    if repo_config_url != None:
        print("loading additional configuration from \"{}\"".format(repo_config_url))
        output += config_from(repo_config_url)
    else:
        output += "\n# REPO_CI_CONFIG_URL was not set\n"

    # Add 3)
    output += config_from(".cirrus.tasks.yml")

    return output


def config_from(config_src):
    """return contents of config file `config_src`, surrounded by markers
    indicating start / end of the the included file
    """

    config_contents = fs.read(config_src)
    config_fmt = """

###
# contents of config file `{0}` start here
###
{1}
###
# contents of config file `{0}` end here
###
"""
    return config_fmt.format(config_src, config_contents)
