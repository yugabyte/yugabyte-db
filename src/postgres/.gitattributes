*		whitespace=space-before-tab,trailing-space
*.[chly]	whitespace=space-before-tab,trailing-space,indent-with-non-tab,tabwidth=4
*.pl		whitespace=space-before-tab,trailing-space,tabwidth=4
*.pm		whitespace=space-before-tab,trailing-space,tabwidth=4
*.po		whitespace=space-before-tab,trailing-space,tab-in-indent,-blank-at-eof
*.sgml		whitespace=space-before-tab,trailing-space,tab-in-indent
*.x[ms]l	whitespace=space-before-tab,trailing-space,tab-in-indent

# Avoid confusing ASCII underlines with leftover merge conflict markers
README		conflict-marker-size=32
README.*	conflict-marker-size=32

# Certain data files that contain special whitespace, and other special cases
*.data						-whitespace
contrib/pgcrypto/sql/pgp-armor.sql		whitespace=-blank-at-eol
src/backend/catalog/sql_features.txt		whitespace=space-before-tab,blank-at-eof,-blank-at-eol
src/backend/utils/Gen_dummy_probes.pl.prolog	whitespace=-blank-at-eof

# Test output files that contain extra whitespace
*.out					-whitespace
src/interfaces/ecpg/test/expected/*	-whitespace

# These files are maintained or generated elsewhere.  We take them as is.
configure				-whitespace
ppport.h				-whitespace
src/backend/jit/llvm/SectionMemoryManager.cpp		-whitespace
src/backend/jit/llvm/SectionMemoryManager.LICENSE	-whitespace
src/backend/regex/COPYRIGHT		-whitespace
src/backend/snowball/libstemmer/*.c	-whitespace
src/backend/utils/mb/Unicode/*-std.txt	-whitespace
src/include/jit/SectionMemoryManager.h	-whitespace
src/include/snowball/libstemmer/*	-whitespace
src/timezone/data/*			-whitespace
