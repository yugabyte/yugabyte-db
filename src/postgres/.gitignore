# Global excludes across all subdirectories
*.o
*.obj
*.bc
*.so
*.so.[0-9]
*.so.[0-9].[0-9]
*.so.[0-9].[0-9][0-9]
*.sl
*.sl.[0-9]
*.sl.[0-9].[0-9]
*.sl.[0-9].[0-9][0-9]
*.dylib
*.dll
*.exp
*.a
*.mo
*.pot
objfiles.txt
.deps/
*.gcno
*.gcda
*.gcov
*.gcov.out
lcov*.info
coverage/
coverage-html-stamp
*.vcproj
*.vcxproj
win32ver.rc
*.exe
lib*dll.def
lib*.pc

# Local excludes in root directory
/GNUmakefile
/config.cache
/config.log
/config.status
/pgsql.sln
/pgsql.sln.cache
/Debug/
/Release/
/tmp_install/
/portlock/
