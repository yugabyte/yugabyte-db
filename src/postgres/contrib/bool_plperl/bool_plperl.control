# bool_plperl extension
comment = 'transform between bool and plperl'
default_version = '1.0'
module_pathname = '$libdir/bool_plperl'
relocatable = true
trusted = true
requires = 'plperl'
