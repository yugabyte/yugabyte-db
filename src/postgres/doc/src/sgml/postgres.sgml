<!-- doc/src/sgml/postgres.sgml -->

<!DOCTYPE book PUBLIC "-//OASIS//DTD DocBook XML V4.5//EN"
          "http://www.oasis-open.org/docbook/xml/4.5/docbookx.dtd"
[

<!ENTITY % version SYSTEM "version.sgml">
%version;
<!ENTITY % filelist SYSTEM "filelist.sgml">
%filelist;

<!ENTITY reference  SYSTEM "reference.sgml">

<!--
Zero-width space.  Use this to allow line breaks at desirable places in
table cells, examples, etc. without causing an unwanted space when the
break is not needed in a wider output rendering.
-->
<!ENTITY zwsp "&#x200B;">

]>

<book id="postgres">
 <title>PostgreSQL &version; Documentation</title>

 <bookinfo>
  <corpauthor>The PostgreSQL Global Development Group</corpauthor>
  <productname>PostgreSQL</productname>
  <productnumber>&version;</productnumber>
  &legal;
 </bookinfo>

 &intro;

 <part id="tutorial">
  <title>Tutorial</title>

  <partintro>
   <para>
    Welcome to the <productname>PostgreSQL</productname> Tutorial.  The
    following few chapters are intended to give a simple introduction
    to <productname>PostgreSQL</productname>, relational database
    concepts, and the SQL language to those who are new to any one of
    these aspects.  We only assume some general knowledge about how to
    use computers.  No particular Unix or programming experience is
    required.  This part is mainly intended to give you some hands-on
    experience with important aspects of the
    <productname>PostgreSQL</productname> system.  It makes no attempt
    to be a complete or thorough treatment of the topics it covers.
   </para>

   <para>
    After you have worked through this tutorial you might want to move
    on to reading <xref linkend="sql"/> to gain a more formal knowledge
    of the SQL language, or <xref linkend="client-interfaces"/> for
    information about developing applications for
    <productname>PostgreSQL</productname>.  Those who set up and
    manage their own server should also read <xref linkend="admin"/>.
   </para>
  </partintro>

  &start;
  &query;
  &advanced;

 </part>

 <part id="sql">
  <title>The SQL Language</title>

  <partintro>
   <para>
    This part describes the use of the <acronym>SQL</acronym> language
    in <productname>PostgreSQL</productname>.  We start with
    describing the general syntax of <acronym>SQL</acronym>, then
    explain how to create the structures to hold data, how to populate
    the database, and how to query it.  The middle part lists the
    available data types and functions for use in
    <acronym>SQL</acronym> commands.  The rest treats several
    aspects that are important for tuning a database for optimal
    performance.
   </para>

   <para>
    The information in this part is arranged so that a novice user can
    follow it start to end to gain a full understanding of the topics
    without having to refer forward too many times.  The chapters are
    intended to be self-contained, so that advanced users can read the
    chapters individually as they choose.  The information in this
    part is presented in a narrative fashion in topical units.
    Readers looking for a complete description of a particular command
    should see <xref linkend="reference"/>.
   </para>

   <para>
    Readers of this part should know how to connect to a
    <productname>PostgreSQL</productname> database and issue
    <acronym>SQL</acronym> commands.  Readers that are unfamiliar with
    these issues are encouraged to read <xref linkend="tutorial"/>
    first.  <acronym>SQL</acronym> commands are typically entered
    using the <productname>PostgreSQL</productname> interactive terminal
    <application>psql</application>, but other programs that have
    similar functionality can be used as well.
   </para>
  </partintro>

  &syntax;
  &ddl;
  &dml;
  &queries;
  &datatype;
  &func;
  &typeconv;
  &indices;
  &textsearch;
  &mvcc;
  &perform;
  &parallel;

 </part>

 <part id="admin">
  <title>Server Administration</title>

  <partintro>
   <para>
    This part covers topics that are of interest to a
    <productname>PostgreSQL</productname> database administrator.  This includes
    installation of the software, set up and configuration of the
    server, management of users and databases, and maintenance tasks.
    Anyone who runs a <productname>PostgreSQL</productname> server, even for
    personal use, but especially in production, should be familiar
    with the topics covered in this part.
   </para>

   <para>
    The information in this part is arranged approximately in the
    order in which a new user should read it.  But the chapters are
    self-contained and can be read individually as desired.  The
    information in this part is presented in a narrative fashion in
    topical units.  Readers looking for a complete description of a
    particular command should see <xref linkend="reference"/>.
   </para>

   <para>
    The first few chapters are written so they can be understood
    without prerequisite knowledge, so new users who need to set
    up their own server can begin their exploration with this part.
    The rest of this part is about tuning and management; that material
    assumes that the reader is familiar with the general use of
    the <productname>PostgreSQL</productname> database system.  Readers are
    encouraged to look at <xref linkend="tutorial"/> and <xref
    linkend="sql"/> for additional information.
   </para>
  </partintro>

  &installbin;
  &installation;
  &installw;
  &runtime;
  &config;
  &client-auth;
  &user-manag;
  &manage-ag;
  &charset;
  &maintenance;
  &backup;
  &high-availability;
  &monitoring;
  &diskusage;
  &wal;
  &logical-replication;
  &jit;
  &regress;

 </part>

 <part id="client-interfaces">
  <title>Client Interfaces</title>

  <partintro>
   <para>
    This part describes the client programming interfaces distributed
    with <productname>PostgreSQL</productname>.  Each of these chapters can be
    read independently.  Note that there are many other programming
    interfaces for client programs that are distributed separately and
    contain their own documentation (<xref linkend="external-projects"/>
    lists some of the more popular ones).  Readers of this part should be
    familiar with using <acronym>SQL</acronym> commands to manipulate
    and query the database (see <xref linkend="sql"/>) and of course
    with the programming language that the interface uses.
   </para>
  </partintro>

  &libpq;
  &lobj;
  &ecpg;
  &infoschema;

 </part>

 <part id="server-programming">
  <title>Server Programming</title>

  <partintro>
   <para>
    This part is about extending the server functionality with
    user-defined functions, data types, triggers, etc.  These are
    advanced topics which should probably be approached only after all
    the other user documentation about <productname>PostgreSQL</productname> has
    been understood.  Later chapters in this part describe the server-side
    programming languages available in the
    <productname>PostgreSQL</productname> distribution as well as
    general issues concerning server-side programming languages.  It
    is essential to read at least the earlier sections of <xref
    linkend="extend"/> (covering functions) before diving into the
    material about server-side programming languages.
   </para>
  </partintro>

  &extend;
  &trigger;
  &event-trigger;
  &rules;

  &xplang;
  &plsql;
  &pltcl;
  &plperl;
  &plpython;

  &spi;
  &bgworker;
  &logicaldecoding;
  &replication-origins;
  &archive-modules;

 </part>

 &reference;

 <part id="internals">
  <title>Internals</title>

  <partintro>
   <para>
    This part contains assorted information that might be of use to
    <productname>PostgreSQL</productname> developers.
   </para>
  </partintro>

  &arch-dev;
  &catalogs;
  &system-views;
  &protocol;
  &sources;
  &nls;
  &plhandler;
  &fdwhandler;
  &tablesample-method;
  &custom-scan;
  &geqo;
  &tableam;
  &indexam;
  &generic-wal;
  &custom-rmgr;
  &btree;
  &gist;
  &spgist;
  &gin;
  &brin;
  &hash;
  &storage;
  &bki;
  &planstats;
  &backup-manifest;

 </part>

 <part id="appendixes">
  <title>Appendixes</title>

  &errcodes;
  &datetime;
  &keywords;
  &features;
  &release;
  &contrib;
  &external-projects;
  &sourcerepo;
  &docguide;
  &limits;
  &acronyms;
  &glossary;
  &color;
  &obsolete;

 </part>

 &biblio;
 <index id="bookindex"></index>

</book>
