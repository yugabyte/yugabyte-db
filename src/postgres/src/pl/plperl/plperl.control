# plperl extension
comment = 'PL/Perl procedural language'
default_version = '1.0'
module_pathname = '$libdir/plperl'
relocatable = false
schema = pg_catalog
superuser = true
trusted = true
