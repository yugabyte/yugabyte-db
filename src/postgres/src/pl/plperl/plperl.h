/*-------------------------------------------------------------------------
 *
 * plperl.h
 *	  Common include file for PL/Perl files
 *
 * This should be included _AFTER_ postgres.h and system include files
 *
 * Portions Copyright (c) 1996-2022, PostgreSQL Global Development Group
 * Portions Copyright (c) 1995, Regents of the University of California
 *
 * src/pl/plperl/plperl.h
 */

#ifndef PL_PERL_H
#define PL_PERL_H

/* stop perl headers from hijacking stdio and other stuff on Windows */
#ifdef WIN32
#define WIN32IO_IS_STDIO
#endif							/* WIN32 */

/*
 * Supply a value of PERL_UNUSED_DECL that will satisfy gcc - the one
 * perl itself supplies doesn't seem to.
 */
#define PERL_UNUSED_DECL pg_attribute_unused()

/*
 * Sometimes perl carefully scribbles on our *printf macros.
 * So we undefine them here and redefine them after it's done its dirty deed.
 */
#undef vsnprintf
#undef snprintf
#undef vsprintf
#undef sprintf
#undef vfprintf
#undef fprintf
#undef vprintf
#undef printf

/*
 * Perl scribbles on the "_" macro too.
 */
#undef _

/*
 * ActivePerl 5.18 and later are MinGW-built, and their headers use GCC's
 * __inline__.  Translate to something MSVC recognizes. Also, perl.h sometimes
 * defines isnan, so undefine it here and put back the definition later if
 * perl.h doesn't.
 */
#ifdef _MSC_VER
#define __inline__ inline
#ifdef isnan
#undef isnan
#endif
/* Work around for using MSVC and Strawberry Perl >= 5.30. */
#define __builtin_expect(expr, val) (expr)
#endif

/*
 * Regarding bool, both PostgreSQL and Perl might use stdbool.h or not,
 * depending on configuration.  If both agree, things are relatively harmless.
 * If not, things get tricky.  If PostgreSQL does but Perl does not, define
 * HAS_BOOL here so that Perl does not redefine bool; this avoids compiler
 * warnings.  If PostgreSQL does not but Perl does, we need to undefine bool
 * after we include the Perl headers; see below.
 */
#ifdef PG_USE_STDBOOL
#define HAS_BOOL 1
#endif

/*
 * Newer versions of the perl headers trigger a lot of warnings with our
 * compiler flags (at least -Wdeclaration-after-statement,
 * -Wshadow=compatible-local are known to be problematic). The system_header
 * pragma hides warnings from within the rest of this file, if supported.
 */
#ifdef HAVE_PRAGMA_GCC_SYSTEM_HEADER
#pragma GCC system_header
#endif

/*
 * Get the basic Perl API.  We use PERL_NO_GET_CONTEXT mode so that our code
 * can compile against MULTIPLICITY Perl builds without including XSUB.h.
 */
#define PERL_NO_GET_CONTEXT
#include "EXTERN.h"
#include "perl.h"

/*
 * We want to include XSUB.h only within .xs files, because on some platforms
 * it undesirably redefines a lot of libc functions.  But it must appear
 * before ppport.h, so use a #define flag to control inclusion here.
 */
#ifdef PG_NEED_PERL_XSUB_H
/*
 * On Windows, win32_port.h defines macros for a lot of these same functions.
 * To avoid compiler warnings when XSUB.h redefines them, #undef our versions.
 */
#ifdef WIN32
#undef accept
#undef bind
#undef connect
#undef fopen
#undef fstat
#undef kill
#undef listen
#undef lstat
#undef mkdir
#undef open
#undef putenv
#undef recv
#undef rename
#undef select
#undef send
#undef socket
#undef stat
#undef unlink
#endif

#include "XSUB.h"
#endif

/* put back our *printf macros ... this must match src/include/port.h */
#ifdef vsnprintf
#undef vsnprintf
#endif
#ifdef snprintf
#undef snprintf
#endif
#ifdef vsprintf
#undef vsprintf
#endif
#ifdef sprintf
#undef sprintf
#endif
#ifdef vfprintf
#undef vfprintf
#endif
#ifdef fprintf
#undef fprintf
#endif
#ifdef vprintf
#undef vprintf
#endif
#ifdef printf
#undef printf
#endif

#define vsnprintf		pg_vsnprintf
#define snprintf		pg_snprintf
#define vsprintf		pg_vsprintf
#define sprintf			pg_sprintf
#define vfprintf		pg_vfprintf
#define fprintf			pg_fprintf
#define vprintf			pg_vprintf
#define printf(...)		pg_printf(__VA_ARGS__)

/*
 * Put back "_" too; but rather than making it just gettext() as the core
 * code does, make it dgettext() so that the right things will happen in
 * loadable modules (if they've set up TEXTDOMAIN correctly).  Note that
 * we can't just set TEXTDOMAIN here, because this file is used by more
 * extensions than just PL/Perl itself.
 */
#undef _
#define _(x) dgettext(TEXTDOMAIN, x)

/* put back the definition of isnan if needed */
#ifdef _MSC_VER
#ifndef isnan
#define isnan(x) _isnan(x)
#endif
#endif

/* perl version and platform portability */
#include "ppport.h"

/*
 * perl might have included stdbool.h.  If we also did that earlier (see c.h),
 * then that's fine.  If not, we probably rejected it for some reason.  In
 * that case, undef bool and proceed with our own bool.  (Note that stdbool.h
 * makes bool a macro, but our own replacement is a typedef, so the undef
 * makes ours visible again).
 */
#ifndef PG_USE_STDBOOL
#ifdef bool
#undef bool
#endif
#endif

/* supply HeUTF8 if it's missing - ppport.h doesn't supply it, unfortunately */
#ifndef HeUTF8
#define HeUTF8(he)			   ((HeKLEN(he) == HEf_SVKEY) ?			   \
								SvUTF8(HeKEY_sv(he)) :				   \
								(U32)HeKUTF8(he))
#endif

/* supply GvCV_set if it's missing - ppport.h doesn't supply it, unfortunately */
#ifndef GvCV_set
#define GvCV_set(gv, cv)		(GvCV(gv) = cv)
#endif

/* Perl 5.19.4 changed array indices from I32 to SSize_t */
#if PERL_BCDVERSION >= 0x5019004
#define AV_SIZE_MAX SSize_t_MAX
#else
#define AV_SIZE_MAX I32_MAX
#endif

/* declare routines from plperl.c for access by .xs files */
HV		   *plperl_spi_exec(char *, int);
void		plperl_return_next(SV *);
SV		   *plperl_spi_query(char *);
SV		   *plperl_spi_fetchrow(char *);
SV		   *plperl_spi_prepare(char *, int, SV **);
HV		   *plperl_spi_exec_prepared(char *, HV *, int, SV **);
SV		   *plperl_spi_query_prepared(char *, int, SV **);
void		plperl_spi_freeplan(char *);
void		plperl_spi_cursor_close(char *);
void		plperl_spi_commit(void);
void		plperl_spi_rollback(void);
char	   *plperl_sv_to_literal(SV *, char *);
void		plperl_util_elog(int level, SV *msg);

#endif							/* PL_PERL_H */
