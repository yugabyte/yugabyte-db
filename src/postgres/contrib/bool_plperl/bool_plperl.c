#include "postgres.h"

#include "fmgr.h"
#include "plperl.h"


PG_MODULE_MAGIC;

PG_FUNCTION_INFO_V1(bool_to_plperl);

Datum
bool_to_plperl(PG_FUNCTION_ARGS)
{
	dTHX;
	bool		in = PG_GETARG_BOOL(0);

	return PointerGetDatum(in ? &PL_sv_yes : &PL_sv_no);
}


PG_FUNCTION_INFO_V1(plperl_to_bool);

Datum
plperl_to_bool(PG_FUNCTION_ARGS)
{
	dTHX;
	SV		   *in = (SV *) PG_GETARG_POINTER(0);

	PG_RETURN_BOOL(SvTRUE(in));
}
