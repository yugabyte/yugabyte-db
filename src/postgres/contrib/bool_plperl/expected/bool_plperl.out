CREATE EXTENSION bool_plperl CASCADE;
NOTICE:  installing required extension "plperl"
--- test transforming from perl
CREATE FUNCTION perl2int(int) RETURNS bool
LANGUAGE plperl
TRANSFORM FOR TYPE bool
AS $$
return shift;
$$;
CREATE FUNCTION perl2text(text) RETURNS bool
LANGUAGE plperl
TRANSFORM FOR TYPE bool
AS $$
return shift;
$$;
CREATE FUNCTION perl2undef() RETURNS bool
LANGUAGE plperl
TRANSFORM FOR TYPE bool
AS $$
return undef;
$$;
SELECT perl2int(1);
 perl2int 
----------
 t
(1 row)

SELECT perl2int(0);
 perl2int 
----------
 f
(1 row)

SELECT perl2text('foo');
 perl2text 
-----------
 t
(1 row)

SELECT perl2text('');
 perl2text 
-----------
 f
(1 row)

SELECT perl2undef() IS NULL AS p;
 p 
---
 t
(1 row)

--- test transforming to perl
CREATE FUNCTION bool2perl(bool, bool, bool) RETURNS void
LANGUAGE plperl
TRANSFORM FOR TYPE bool, for type boolean  -- duplicate to test ruleutils
AS $$
my ($x, $y, $z) = @_;

die("NULL mistransformed") if (defined($z));
die("TRUE mistransformed to UNDEF") if (!defined($x));
die("FALSE mistransformed to UNDEF") if (!defined($y));
die("TRUE mistransformed") if (!$x);
die("FALSE mistransformed") if ($y);
$$;
SELECT bool2perl (true, false, NULL);
 bool2perl 
-----------
 
(1 row)

--- test ruleutils
\sf bool2perl
CREATE OR REPLACE FUNCTION public.bool2perl(boolean, boolean, boolean)
 RETURNS void
 TRANSFORM FOR TYPE boolean, FOR TYPE boolean
 LANGUAGE plperl
AS $function$
my ($x, $y, $z) = @_;

die("NULL mistransformed") if (defined($z));
die("TRUE mistransformed to UNDEF") if (!defined($x));
die("FALSE mistransformed to UNDEF") if (!defined($y));
die("TRUE mistransformed") if (!$x);
die("FALSE mistransformed") if ($y);
$function$
--- test selecting bool through SPI
CREATE FUNCTION spi_test()  RETURNS void
LANGUAGE plperl
TRANSFORM FOR TYPE bool
AS $$
my $rv = spi_exec_query('SELECT true t, false f, NULL n')->{rows}->[0];

die("TRUE mistransformed to UNDEF in SPI") if (!defined ($rv->{t}));
die("FALSE mistransformed to UNDEF in SPI") if (!defined ($rv->{f}));
die("NULL mistransformed in SPI") if (defined ($rv->{n}));
die("TRUE mistransformed in SPI") if (!$rv->{t});
die("FALSE mistransformed in SPI") if ($rv->{f});
$$;
SELECT spi_test();
 spi_test 
----------
 
(1 row)

DROP EXTENSION plperl CASCADE;
NOTICE:  drop cascades to 6 other objects
DETAIL:  drop cascades to function spi_test()
drop cascades to extension bool_plperl
drop cascades to function perl2int(integer)
drop cascades to function perl2text(text)
drop cascades to function perl2undef()
drop cascades to function bool2perl(boolean,boolean,boolean)
